### PR TITLE
add websocket proxy support

### PR DESCRIPTION
**Pull request by aruis**
Signed-off-by: Liu Rui <lovearuis@gmail.com>

Motivation:

add websocket proxy support [issue14](https://github.com/eclipse-vertx/vertx-http-proxy/issues/14)

### CODE REVIEW

**Comment by vietj** [comment]
can you sign the eclipse contributor agreement ?

**Comment by aruis** [comment]
> can you sign the eclipse contributor agreement ?

Ok, I have handled it. @vietj

File: src/main/java/io/vertx/httpproxy/impl/ProxyRequestImpl.java [hunk 1]
```diff
@@ -193,4 +185,14 @@ void send(HttpClientRequest inboundRequest, Handler<AsyncResult<ProxyResponse>>
     this.inboundRequest = inboundRequest;
     sendRequest(completionHandler);
   }
+
+  void websocket(WebSocket inboundWebSocket) {
+    Future<ServerWebSocket> fut = outboundRequest.toWebSocket();
+    fut.onSuccess(ws -> {
```
Comment (vietj) [inline src/main/java/io/vertx/httpproxy/impl/ProxyRequestImpl.java:191]: we should handle and test the failure case

File: src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java [hunk 2]
```diff
@@ -69,6 +63,14 @@ public void handle(HttpServerRequest outboundRequest) {
     });
   }
 
+  private Future<WebSocket> resolveOriginWS(HttpServerRequest outboundRequest) {
+    return selector.apply(outboundRequest).flatMap(server -> {
+      WebSocketConnectOptions webSocketConnectOptions = new WebSocketConnectOptions();
+      webSocketConnectOptions.setServer(server);
```
Comment (vietj) [inline src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java:69]: we are missing stuff here like transmitting relevant HTTP headers from the client, some headers might contain important information like authentication

File: src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java [hunk 3]
```diff
@@ -69,6 +63,14 @@ public void handle(HttpServerRequest outboundRequest) {
     });
   }
 
+  private Future<WebSocket> resolveOriginWS(HttpServerRequest outboundRequest) {
+    return selector.apply(outboundRequest).flatMap(server -> {
+      WebSocketConnectOptions webSocketConnectOptions = new WebSocketConnectOptions();
```
Comment (vietj) [inline src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java:68]: there is not opportunity to configure WebSocketConnectOptions to the origin.

File: src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java [hunk 4]
```diff
@@ -69,6 +63,14 @@ public void handle(HttpServerRequest outboundRequest) {
     });
   }
 
+  private Future<WebSocket> resolveOriginWS(HttpServerRequest outboundRequest) {
```
Comment (vietj) [inline src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java:66]: this should be renamed as it does not resolve the original WS but instead resolve the origin and then connect to the resolved address

**Review by vietj — CHANGES_REQUESTED** [review]
a few concerns have been done, we should also check how other similar project handle this.

File: src/main/java/io/vertx/httpproxy/impl/ProxyRequestImpl.java (hunk 1 above)
Comment (aruis) [reply src/main/java/io/vertx/httpproxy/impl/ProxyRequestImpl.java]: The relevant code has been improved, but I am not sure how the unit test is written.

File: src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java (hunk 2 above)
Comment (aruis) [reply src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java:69]: Related code has been fixed.

File: src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java (hunk 3 above)
Comment (aruis) [reply src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java:68]: Related code has been fixed.

File: src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java (hunk 4 above)
Comment (aruis) [reply src/main/java/io/vertx/httpproxy/impl/HttpProxyImpl.java]: Related code has been fixed.

**Comment by aruis** [comment]
Thank you for pointing out my question, I tried to fix some code, please review it. @vietj

**Comment by vietj** [comment]
one concern is about the proxy connection to the server and reusing options sent by the client

**Comment by aruis** [comment]
Hello, you propose questions, key information is in the private property `handshaker` of `ServerWebsocketImp`, I don't know what to do.😕 @vietj

**Comment by vietj** [comment]
@aruis yes, so I think we should see what other web socket proxy do with the regards and we should expose this information from the web socket if that is needed for the proxy

**Comment by vietj** [comment]
can you dig to see what other web socket proxy do ?

**Comment by aruis** [comment]
Learn Spring Cloud Gateway's way to work, it seems that they only read and set "SEC-Websocket-Protocol" and then forwarded the other header.

https://github.com/spring-cloud/spring-cloud-gateway/blob/71014d10bfa040a08d3479b161b79b3c8cd0d4a8/spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/WebsocketRoutingFilter.java#L108

https://github.com/spring-projects/spring-framework/blob/b595dc1dfad9db534ca7b9e8f46bb9926b88ab5a/spring-webflux/src/main/java/org/springframework/web/reactive/socket/server/support/HandshakeWebSocketService.java#L286

**Comment by vietj** [comment]
I would not consider spring gateway as a reference in that domain, instead I think we should rather look at what node.js similar tools and classic proxies like nginx or ha-proxy do

**Comment by vietj** [comment]
I don't see any testing of failures, there should be test checking all kind of failures

- at connection time
  - connection error
  - server does not handle WebSocket
- at runtime
  - client disconnects abruptly
  - server disconnects abrupbly

**Comment by nragon** [comment]
Nice work. Following.
Btw, I'm using this on a use case where I need websockets. Seems to be working fine.

**Comment by vietj** [comment]
Implemented using handshake forwarding like Nginx does : https://github.com/eclipse-vertx/vertx-http-proxy/pull/27

**Pull request closed without merging**